jcl/src/java.base/share/classes/java/lang/String.java

```diff
@@ -1794,29 +1794,32 @@ public boolean equalsIgnoreCase(String string) {
 			byte byteAtO1Last = helpers.getByteFromArrayByIndex(s1Value, s1len - 1);
 			byte byteAtO2Last = helpers.getByteFromArrayByIndex(s2Value, s1len - 1);
 
-			if (byteAtO1Last != byteAtO2Last &&
-					toUpperCase(helpers.byteToCharUnsigned(byteAtO1Last)) != toUpperCase(helpers.byteToCharUnsigned(byteAtO2Last))) {
+			if ((byteAtO1Last != byteAtO2Last)
+					&& (toUpperCase(helpers.byteToCharUnsigned(byteAtO1Last)) != toUpperCase(helpers.byteToCharUnsigned(byteAtO2Last)))
+			) {
 				return false;
 			}
 
 			while (o1 < end - 1) {
 				byte byteAtO1 = helpers.getByteFromArrayByIndex(s1Value, o1++);
 				byte byteAtO2 = helpers.getByteFromArrayByIndex(s2Value, o2++);
 
-				if (byteAtO1 != byteAtO2 &&
-						toUpperCase(helpers.byteToCharUnsigned(byteAtO1)) != toUpperCase(helpers.byteToCharUnsigned(byteAtO2))) {
+				if ((byteAtO1 != byteAtO2)
+						&& (toUpperCase(helpers.byteToCharUnsigned(byteAtO1)) != toUpperCase(helpers.byteToCharUnsigned(byteAtO2)))
+				) {
 					return false;
 				}
 			}
 		} else {
-			// Compare the last chars. Under string compression, the compressible char set obeys 1-1 mapping for upper/lower
-			// case, converting to lower cases then compare should be sufficient.
+			// Compare the last chars.
+			// In order to tell 2 chars are different:
+			// If at least one char is ASCII, converting to upper cases then compare should be sufficient.
+			// If both chars are not in ASCII char set, need to convert to lower case and compare as well.
 			char charAtO1Last = s1.charAtInternal(s1len - 1, s1Value);
 			char charAtO2Last = s2.charAtInternal(s1len - 1, s2Value);
 
-			if (charAtO1Last != charAtO2Last
-					&& toUpperCase(charAtO1Last) != toUpperCase(charAtO2Last)
-					&& ((charAtO1Last <= 255 && charAtO2Last <= 255) || Character.toLowerCase(charAtO1Last) != Character.toLowerCase(charAtO2Last))
```

Comment:
@fjeremic fyi. We're losing the <= 256 optimization here, do you think it's worth putting this into compareValue()?

Reply:
I guess that doesn't work as is, we've have to create a new helper function that takes both chars.

Reply:
Yeah should be fine as the check has moved into `charValuesEqualIgnoreCase`.